### PR TITLE
FreeBSD build support

### PR DESCRIPTION
**Pull request by asdfugil**
[HOWTO](https://github.com/asdfugil/Procursus/blob/freebsd/FREEBSD.md)

### CODE REVIEW

File: FREEBSD.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Building on FreeBSD
+Building on FreeBSD is made possible by cctools-port.
+
+This assumes you have sudo installed and configured.
+
+1. install packages
+```
+sudo pkg install findutils llvm12 git gmake patch gnugrep po4a docbook-xsl ncurses autoconf automake gettext libtool pkgconf dpkg fakeroot zstd python39 coreutils bash cmake gnupg openssl gsed gtar perl5 wget
```
Comment (CRKatri) [inline FREEBSD.md:8]: Is there a reason you want llvm12 instead of using the clang in world?

File: Makefile [hunk 2]
```diff
@@ -193,6 +193,25 @@ OTOOL    := $(GNU_HOST_TRIPLE)-otool
...
+LD              := $(GNU_HOST_TRIPLE)-ld 
+RANLIB    := $(GNU_HOST_TRIPLE)-ranlib   
+STRIP       := $(GNU_HOST_TRIPLE)-strip
+I_N_T       := $(GNU_HOST_TRIPLE)-install_name_tool
+NM            := $(GNU_HOST_TRIPLE)-nm
+LIPO         := $(GNU_HOST_TRIPLE)-lipo
+OTOOL    := $(GNU_HOST_TRIPLE)-otool
+EXTRA      := INSTALL="/usr/bin/install -c --strip-program=$(STRIP)"
+LIBTOOL := $(GNU_HOST_TRIPLE)-libtool
+PATH        := $(BUILD_DIR)/gnubin:$(BUILD_DIR)/cctools-src/usage_examples/ios_toolchain/target/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin:/rescue
```
Comment (CRKatri) [inline Makefile:213]: Prepend to PATH instead of overwriting.

File: Makefile [hunk 3]
```diff
@@ -193,6 +193,25 @@ OTOOL    := $(GNU_HOST_TRIPLE)-otool
...
+AR              := $(GNU_HOST_TRIPLE)-ar
+LD              := $(GNU_HOST_TRIPLE)-ld 
+RANLIB    := $(GNU_HOST_TRIPLE)-ranlib   
+STRIP       := $(GNU_HOST_TRIPLE)-strip
+I_N_T       := $(GNU_HOST_TRIPLE)-install_name_tool
+NM            := $(GNU_HOST_TRIPLE)-nm
+LIPO         := $(GNU_HOST_TRIPLE)-lipo
+OTOOL    := $(GNU_HOST_TRIPLE)-otool
+EXTRA      := INSTALL="/usr/bin/install -c --strip-program=$(STRIP)"
+LIBTOOL := $(GNU_HOST_TRIPLE)-libtool
```
Comment (CRKatri) [inline Makefile:212]: Fix all this alignment

**Review by CRKatri — CHANGES_REQUESTED** [review]

File: Makefile [hunk 4]
```diff
@@ -193,6 +193,25 @@ OTOOL    := $(GNU_HOST_TRIPLE)-otool
 EXTRA    := INSTALL="/usr/bin/install -c --strip-program=$(STRIP)"
 LIBTOOL  := $(GNU_HOST_TRIPLE)-libtool
 
+else ifeq ($(UNAME),FreeBSD)
+$(warning Building on FreeBSD)
```
Comment (CRKatri) [inline Makefile:197]: Add the `MEMO_QUIET` wrapper.

**Review by CRKatri — CHANGES_REQUESTED** [review]

**Comment by CRKatri** [comment]
Thanks for this, I'm gonna have to setup `gnubin` and give it a shot, and maybe look into adding `gnubin` to the port.

File: FREEBSD.md (hunk 1 above)
Comment (asdfugil) [reply FREEBSD.md:8]: because clang doesn't exist

Comment (CRKatri) [reply FREEBSD.md:8]: wdym? clang is part of world, https://cgit.freebsd.org/src/tree/usr.bin/clang

**Comment by asdfugil** [comment]
> Thanks for this, I'm gonna have to setup `gnubin` and give it a shot, and maybe look into adding `gnubin` to the port.

that script sets it up for you btw

**Comment by CRKatri** [comment]
I know, but I already have cctools and my SDKs setup.

File: FREEBSD.md (hunk 1 above)
Comment (asdfugil) [reply FREEBSD.md:8]:
```
[nick@nick-pc ~]$ pkg search clang
flang-clang-7.0.g20191020      Fortran compiler targeting LLVM (modified clang part)
opencl-clang-11.0.0            Clang wrapper to compile OpenCL C kernels to SPIR-V modules
[nick@nick-pc ~]$ 
```

File: Makefile (hunk 3 above)
Comment (asdfugil) [reply Makefile:212]: lmao vscode breaks alignment gonna use nano

File: FREEBSD.md (hunk 1 above)
Comment (CRKatri) [reply FREEBSD.md:8]: world isn't tracked by pkg, run `clang --version`

Comment (asdfugil) [reply FREEBSD.md:8]:
```
[nick@nick-pc ~]$ clang --version
FreeBSD clang version 11.0.1 (git@github.com:llvm/llvm-project.git llvmorg-11.0.1-0-g43ff75f2c3fe)
Target: x86_64-unknown-freebsd13.0
Thread model: posix
InstalledDir: /usr/bin
```
so i guess i should remove llvm12?

**Comment by asdfugil** [comment]
all right fixed

**Comment by CRKatri** [comment]
Does `gmake libxcrypt` work for you? If it does I'll approve this PR and try to fix my setup, otherwise we gotta look into that.

**Comment by asdfugil** [comment]
> Does `gmake libxcrypt` work for you? If it does I'll approve this PR and try to fix my setup, otherwise we gotta look into that.
:(
```
checking dependency style of aarch64-apple-darwin-clang... none
checking for aarch64-apple-darwin-pkg-config... no
checking for pkg-config... /usr/local/bin/pkg-config
configure: WARNING: using cross tools not prefixed with host triplet
checking pkg-config is at least version 0.9.0... yes
configure: error: a C99- or C11-compliant compiler is required
gmake: *** [libxcrypt.mk:21: libxcrypt] Error 1
```

**Comment by CRKatri** [comment]
Ok, the issue is that iOS uses llvm for some of the headers, namely `stdarg.h` and `stdbool.h` (but there are some others) and on macOS, iOS and Linux installing llvm installs those headers, but they are part of FreeBSD so neither world or any of the ports contain them. Some of them you can copy over from `Kernel.framework` in the macOS SDK, which we may have to do, unless you have a better idea.

**Comment by asdfugil** [comment]
it works if I copy it
I will add the copying to Makefile

**Comment by asdfugil** [comment]
all right hopefully I added the copying in the correct place.

**Comment by asdfugil** [comment]
the commit timestamps are messed up.

for whatever reason bash needs gcc to build so I have it as a package that has to be installed.

**Review by CRKatri — APPROVED** [review]

**Comment by asdfugil** [comment]
`make bash` is broken again because libiosexec doesn't build. However ,same problem happens on Linux so this is not a FreeBSD-specific problem
```
libtool -static -o libiosexec.a *.o
Usage: /usr/local/bin/libtool [OPTION]... [MODE-ARG]...
Try 'libtool --help' for more information.
libtool:   error: unrecognised option: '-static'
make[1]: *** [Makefile:10: static] Error 1
make: *** [libiosexec.mk:23: libiosexec] Error 2
```

**Comment by CRKatri** [comment]
That has been fixed in upstream `libiosexec` and shouldn't be worried about.

**Comment by asdfugil** [comment]
merge conflict moment

**Comment by asdfugil** [comment]
`make bash-package` should work now.

**Pull request merged**